src/Data/List/Properties.agda

```diff
@@ -122,6 +122,15 @@ map-injective finj {x ∷ xs} {y ∷ ys} eq =
 ------------------------------------------------------------------------
 -- mapMaybe
 
+module _ {f g : A → Maybe B} where
+
+  mapMaybe-cong : f ≗ g → mapMaybe f ≗ mapMaybe g
+  mapMaybe-cong f≗g [] = refl
+  mapMaybe-cong f≗g (x ∷ xs) rewrite f≗g x
+    with ih ← mapMaybe-cong f≗g xs | g x
+  ... | just y  = cong (y ∷_) ih
+  ... | nothing = ih
```

Comment:
Having said all of that, my understand of @JacquesCarette 's point in #2359 is that such proof patterns might be (better?) finessed using the *dependent* function `Maybe.maybe`, in which case the `ih` terms might end up being able to be `let`-bound after all...?

Reply:
The thing with `with`, as I understand it, is that proving things _about_ functions that use `with` can be inconvenient. We only very very rarely prove things about this kind of property. I think `with` is fine throughout this PR

Reply:
My main 'beef' with `with` is when it is used to define functions, as that forces your hand downstream on how your proofs are forced to be done. And taking them out of proofs doesn't give as large a reward.

Having said that, I'm also interested in having proofs be less 'low level' (i.e. forcing explicit evaluation by following the definition patterns of the original function) and having more equational proofs. This would allow a smaller 'core API' that is representation dependent, and a larger set of properties which are representation *independent*.

So it's not so much 'simplification' of a proof as moving towards more intrinsic proofs.

Reply:
note that the use of `with` here mirrors the style of the existing `mapMaybe` proofs. if we want to move away from it, I'd prefer to leave it to a future PR that also changes the definition in Base

Reply:
Absolutely! See #2361 although in the end I did *not* change the proof of `mapMaybe-concatMap`... but maybe I should!?